operator/reconciler.go

```diff
@@ -0,0 +1,276 @@
+			Operations: []resource.PatchOperation{{
+				Operation: resource.PatchOpAdd,
+				Path:      "/metadata/finalizers",
+				Value:     []string{o.finalizer},
+			}},
+		}, resource.PatchOptions{}, request.Object)
+		return ReconcileResult{}, patchErr
+	}
+	return o.wrappedReconcile(ctx, request)
+}
+
+func (o *OpinionatedReconciler) wrappedReconcile(ctx context.Context, request ReconcileRequest) (ReconcileResult, error) {
+	if o.ReconcileFunc != nil {
+		return o.ReconcileFunc(ctx, request)
+	}
+	return ReconcileResult{}, nil
+}
+
+// Wrap wraps the provided Reconciler's Reconcile function with this OpinionatedReconciler
+func (o *OpinionatedReconciler) Wrap(reconciler Reconciler) {
+	o.ReconcileFunc = reconciler.Reconcile
+}
+
+// Compile-time interface compliance check
+var _ Reconciler = &OpinionatedReconciler{}
+
+// SimpleReconciler is a simple Reconciler implementation that calls ReconcileFunc if non-nil on Reconcile requests.
+type SimpleReconciler struct {
+	ReconcileFunc func(context.Context, ReconcileRequest) (ReconcileResult, error)
+}
+
+// Reconcile calls ReconcileFunc if non-nil and returns the response, or returns an empty ReconcileResult and nil error
+// if ReconcileFunc is nil.
+func (s *SimpleReconciler) Reconcile(ctx context.Context, req ReconcileRequest) (ReconcileResult, error) {
+	if s.ReconcileFunc != nil {
+		return s.ReconcileFunc(ctx, req)
+	}
+	return ReconcileResult{}, nil
+}
+
+// Compile-time interface compliance check
+var _ Reconciler = &SimpleReconciler{}
+
+// TypedReconcileRequest is a variation of ReconcileRequest which uses a concretely-typed Object,
+// rather than the interface resource.Object. It is used by TypedReconciler in its ReconcileFunc.
+type TypedReconcileRequest[T resource.Object] struct {
+	// Action is the actions which triggered this TypedReconcileRequest
+	Action ReconcileAction
+	// Object is the object on which the Action was performed, at the point in time of that Action
+	Object T
+	// State is a user-defined map of state values that can be provided on retried ReconcileRequests.
+	// See State in ReconcileResult. It will always be nil on an initial Reconcile call,
+	// and will only be non-nil if a prior Reconcile call with this TypedReconcileRequest returned a State
+	// in its ReconcileResult alongside either a RequeueAfter or an error.
+	State map[string]any
+}
+
+// TypedReconciler is a variant of SimpleReconciler in which a user can specify the underlying type of the resource.Object
+// which is in the provided ReconcileRequest. Reconcile() will then attempt to cast the resource.Object in the
+// ReconcileRequest into the provided T type and produce a TypedReconcileRequest, which will be passed to ReconcileFunc.
+type TypedReconciler[T resource.Object] struct {
+	// ReconcileFunc is called by TypedReconciler.Reconcile using the T-typed Object instead of a resource.Object.
+	ReconcileFunc func(context.Context, TypedReconcileRequest[T]) (ReconcileResult, error)
```

Comment:
Again – I think we should wrap the interface, not a function.

Reply:
We can't wrap the interface here, as we're passing a`TypedReconcileRequest`, rather than a `ReconcileRequest` that uses `resource.Object`, which is what the `Reconciler` interface expects.

Reply:
Ah, good point – I've missed that. Let's keep it the way it is then 👍